homeassistant/components/media_player/cast.py

```diff
@@ -33,92 +39,172 @@
     SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_PREVIOUS_TRACK | \
     SUPPORT_NEXT_TRACK | SUPPORT_PLAY_MEDIA | SUPPORT_STOP | SUPPORT_PLAY
 
-KNOWN_HOSTS_KEY = 'cast_known_hosts'
+INTERNAL_DISCOVERY_RUNNING_KEY = 'cast_discovery_running'
+# UUID -> CastDevice mapping; cast devices without UUID are not stored
+ADDED_CAST_DEVICES_KEY = 'cast_added_cast_devices'
+# Stores every discovered (host, port, uuid)
+KNOWN_CHROMECASTS_KEY = 'cast_all_chromecasts'
+
+SIGNAL_CAST_DISCOVERED = 'cast_discovered'
 
 PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
     vol.Optional(CONF_HOST): cv.string,
     vol.Optional(CONF_IGNORE_CEC): [cv.string],
 })
 
 
-# pylint: disable=unused-argument
-def setup_platform(hass, config, add_devices, discovery_info=None):
-    """Set up the cast platform."""
-    import pychromecast
+def _setup_internal_discovery(hass: HomeAssistantType) -> None:
+    """Set up the pychromecast internal discovery."""
+    hass.data.setdefault(INTERNAL_DISCOVERY_RUNNING_KEY, threading.Lock())
+    if not hass.data[INTERNAL_DISCOVERY_RUNNING_KEY].acquire(blocking=False):
+        # Internal discovery is already running
+        return
 
-    # Import CEC IGNORE attributes
-    pychromecast.IGNORE_CEC += config.get(CONF_IGNORE_CEC, [])
+    import pychromecast
 
-    known_hosts = hass.data.get(KNOWN_HOSTS_KEY)
-    if known_hosts is None:
-        known_hosts = hass.data[KNOWN_HOSTS_KEY] = []
+    def internal_callback(name):
+        """Called when zeroconf has discovered a new chromecast."""
+        mdns = listener.services[name]
+        ip_address, port, uuid, _, _ = mdns
+        key = (ip_address, port, uuid)
 
-    if discovery_info:
-        host = (discovery_info.get('host'), discovery_info.get('port'))
+        if key in hass.data[KNOWN_CHROMECASTS_KEY]:
+            _LOGGER.debug("Discovered previous chromecast %s", mdns)
+            return
 
-        if host in known_hosts:
+        _LOGGER.debug("Discovered new chromecast %s", mdns)
+        try:
+            # pylint: disable=protected-access
+            chromecast = pychromecast._get_chromecast_from_host(
```

Comment:
I'm not entirely happy with having to create the "complete" Chromecast component every time we find some new cast device with discovery. Each call creates a separate thread, socket, ...

It's still much better than the previous version, which would create an entirely new object for **each call of `setup_platform`**, but I think we could just store the mdns info in `hass.data[KNOWN_CHROMECASTS_KEY]` and lazily create chromecast objects inside `CastDevice`. That also make handling disconnects cleaner. But I think that should be in a separate PR (maybe together with `add_reader`).